SignalMessaging/contacts/OWSContactsManager.m

```diff
@@ -225,29 +234,142 @@ - (void)systemContactsFetcher:(SystemContactsFetcher *)systemContactsFetcher
+                // "full" intersection.
+                NSSet<NSString *> *_Nullable lastKnownContactPhoneNumbers =
+                    [transaction objectForKey:OWSContactsManagerKeyLastKnownContactPhoneNumbers
+                                 inCollection:OWSContactsManagerCollection];
+                if (lastKnownContactPhoneNumbers) {
+                    // Do a "delta" sync which only intersects recipient ids not included
+                    // in the last full intersection.
+                    NSMutableSet<NSString *> *newRecipientIds = [allContactRecipientIds mutableCopy];
+                    [newRecipientIds minusSet:lastKnownContactPhoneNumbers];
+                    recipientIdsForIntersection = newRecipientIds;
+                } else {
+                    // Without a list of "last known" contact phone numbers, we'll have to do a full intersection.
+                    isFullIntersection = YES;
+                }
+            }
+        }];
+        OWSAssert(recipientIdsForIntersection);
+
+        if (recipientIdsForIntersection.count < 1) {
+            DDLogInfo(@"%@ Skipping intersection; no contacts to intersect.", self.logTag);
+            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                completion(nil);
+            });
+            return;
+        } else if (isFullIntersection) {
+            DDLogInfo(@"%@ Doing full intersection with %zd contacts.", self.logTag, recipientIdsForIntersection.count);
+        } else {
+            DDLogInfo(
+                @"%@ Doing delta intersection with %zd contacts.", self.logTag, recipientIdsForIntersection.count);
+        }
+
+        [self intersectContacts:recipientIdsForIntersection
+            retryDelaySeconds:1.0
+            success:^(NSSet<SignalRecipient *> *registeredRecipients) {
+                [self markIntersectionAsComplete:allContactRecipientIds isFullIntersection:isFullIntersection];
+
+                completion(nil);
+            }
+            failure:^(NSError *error) {
+                completion(error);
+            }];
+    });
+}
+
+- (void)markIntersectionAsComplete:(NSSet<NSString *> *)recipientIdsForIntersection
+                isFullIntersection:(BOOL)isFullIntersection
+{
+    OWSAssert(recipientIdsForIntersection.count > 0);
+
+    dispatch_async(self.serialQueue, ^{
+        [self.dbReadConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+            [transaction setObject:recipientIdsForIntersection
+                            forKey:OWSContactsManagerKeyLastKnownContactPhoneNumbers
+                      inCollection:OWSContactsManagerCollection];
+
+            if (isFullIntersection) {
+                // Don't do a full intersection more often than once every 6 hours.
+                const NSTimeInterval kFullIntersectionRate = 6 * kHourInterval;
+                NSDate *nextFullIntersectionDate =
```

Comment:
No biggie, but in case we change this interval later, it probably makes more sense to store the `lastFullIntersectionDate` and compute the `nextFullIntersection` wherever it's used.

Reply:
I went this way at Jeff's suggestion because there's future work here to honor the "next full intersection interval" hint which CDS will eventually include in responses.  But happy to do either way if it turns out to be easier.

Reply:
Ah, that makes sense.